File: src/main/java/ru/citeck/ecos/apps/domain/buildinfo/api/records/BuildInfoRecords.java
package ru.citeck.ecos.apps.domain.buildinfo.api.records;

import lombok.Data;
import lombok.RequiredArgsConstructor;
import org.jetbrains.annotations.NotNull;
import org.springframework.stereotype.Component;
import ru.citeck.ecos.apps.app.domain.buildinfo.dto.BuildInfo;
import ru.citeck.ecos.apps.app.remote.AppStatus;
import ru.citeck.ecos.records2.RecordRef;
import ru.citeck.ecos.records2.graphql.meta.value.EmptyValue;
import ru.citeck.ecos.records2.graphql.meta.value.MetaField;
import ru.citeck.ecos.records2.request.query.RecordsQuery;
import ru.citeck.ecos.records2.request.query.RecordsQueryResult;
import ru.citeck.ecos.records2.source.dao.local.LocalRecordsDao;
import ru.citeck.ecos.records2.source.dao.local.v2.LocalRecordsMetaDao;
import ru.citeck.ecos.records2.source.dao.local.v2.LocalRecordsQueryWithMetaDao;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

@Component
public class BuildInfoRecords extends LocalRecordsDao
    implements LocalRecordsQueryWithMetaDao<Object>, LocalRecordsMetaDao<Object> {

    public static final String ID = "build-info";

    private final Map<String, Record> fullBuildInfo = new ConcurrentHashMap<>();

    @Override
    public RecordsQueryResult<Object> queryLocalRecords(@NotNull RecordsQuery recordsQuery,
                                                        @NotNull MetaField metaField) {

        return new RecordsQueryResult<>(new ArrayList<>(fullBuildInfo.values()));
    }

    @Override
    public List<Object> getLocalRecordsMeta(@NotNull List<RecordRef> records,
                                            @NotNull MetaField metaField) {

        return records.stream().map( ref -> {
            Record rec = fullBuildInfo.get(ref.getId());
            if (rec != null) {
                return rec;
            } else {
                return EmptyValue.INSTANCE;
            }
        }).collect(Collectors.toList());
    }

    public void register(AppStatus app, List<BuildInfo> buildInfo) {
        for (BuildInfo info : buildInfo) {
            String id = app.getAppName() + "-" + info.getRepo();
            Record currentInfo = fullBuildInfo.get(id);
            if (currentInfo == null || currentInfo.info.getBuildDate().isBefore(info.getBuildDate())) {
                fullBuildInfo.put(id, new Record(id, app, info));
            }
        }
    }

    @Override
    public String getId() {
        return ID;
    }

    @Data
    @RequiredArgsConstructor
    public static class Record {
        private final String id;
        private final AppStatus app;
        private final BuildInfo info;
    }
}

File: src/main/java/ru/citeck/ecos/apps/domain/buildinfo/api/records/BuildInfoRecords.kt
package ru.citeck.ecos.apps.domain.buildinfo.api.records

import ru.citeck.ecos.records2.source.dao.local.LocalRecordsDao
import ru.citeck.ecos.records2.source.dao.local.v2.LocalRecordsQueryWithMetaDao
import ru.citeck.ecos.records2.source.dao.local.v2.LocalRecordsMetaDao
import java.util.concurrent.ConcurrentHashMap
import ru.citeck.ecos.records2.request.query.RecordsQuery
import ru.citeck.ecos.records2.graphql.meta.value.MetaField
import ru.citeck.ecos.records2.request.query.RecordsQueryResult
import java.util.stream.Collectors
import ru.citeck.ecos.records2.RecordRef
import ru.citeck.ecos.records2.graphql.meta.value.EmptyValue
import org.springframework.stereotype.Component
import ru.citeck.ecos.apps.app.domain.buildinfo.dto.BuildInfo
import ru.citeck.ecos.apps.app.remote.AppStatus
import java.util.*

@Component
class BuildInfoRecords : LocalRecordsDao(), LocalRecordsQueryWithMetaDao<Any>, LocalRecordsMetaDao<Any> {

    companion object {
        const val ID = "build-info"
    }

    private val fullBuildInfo: MutableMap<String, Record> = ConcurrentHashMap()

    override fun queryLocalRecords(recordsQuery: RecordsQuery, field: MetaField): RecordsQueryResult<Any> {
        return RecordsQueryResult(ArrayList<Any>(fullBuildInfo.values))
    }

    override fun getLocalRecordsMeta(records: List<RecordRef>, metaField: MetaField): List<Any> {
        return records.stream().map { ref: RecordRef ->
            fullBuildInfo[ref.id] ?: EmptyValue.INSTANCE
        }.collect(Collectors.toList())
    }

    override fun getId(): String {
        return ID
    }

    fun register(app: AppStatus, buildInfo: List<BuildInfo>) {
        for (info in buildInfo) {
            val id = app.appName + "-" + info.repo
            val currentInfo = fullBuildInfo[id]
            if (currentInfo == null || currentInfo.info.buildDate.isBefore(info.buildDate)) {
                fullBuildInfo[id] = Record(id, app, info)
            }
        }
    }

    class Record(
        val id: String,
        val app: AppStatus,
        val info: BuildInfo
    )
}
